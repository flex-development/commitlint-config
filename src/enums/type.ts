/**
 * @file Enums - Type
 * @module commitlint-config/enums/Type
 */

/**
 * Commit types.
 *
 * @see https://commitlint.js.org/#/reference-rules?id=type-enum
 *
 * @enum {Lowercase<string>}
 */
enum Type {
  BUILD = 'build',
  CHORE = 'chore',
  CI = 'ci',
  DOCS = 'docs',
  FEAT = 'feat',
  FIX = 'fix',
  PERF = 'perf',
  REFACTOR = 'refactor',
  REVERT = 'revert',
  STYLE = 'style',
  TEST = 'test',
  WIP = 'wip'
}

export default Type
