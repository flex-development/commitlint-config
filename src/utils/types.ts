/**
 * @file Utilities - types
 * @module commitlint-config/utils/types
 */

import { Type } from '#src/enums'
import type { LiteralUnion } from '@flex-development/tutils'

/**
 * Returns an array containing valid commit types.
 *
 * @param {(Set<string> | string[])?} [extras=[]] - Additional commit types
 * @return {LiteralUnion<Type, string>[]} Commit types array
 */
const types = (
  extras: Set<string> | string[] = []
): LiteralUnion<Type, string>[] => {
  return [...new Set([...Object.values(Type), ...extras])].sort((t1, t2) => {
    return t1.localeCompare(t2)
  })
}

export default types
