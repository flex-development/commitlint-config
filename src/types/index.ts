/**
 * @file Entry Point - Type Definitions
 * @module commitlint-config/types
 */

export type { default as NoteKeyword } from './note-keyword'
export type { default as ParserPreset } from './parser-preset'
export type { default as Question } from './question'
export type { default as QuestionEnum } from './question-enum'
export type { default as Questions } from './questions'
