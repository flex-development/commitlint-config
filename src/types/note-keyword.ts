/**
 * @file Type Definitions - NoteKeyword
 * @module commitlint-config/types/NoteKeyword
 */

/**
 * Keywords for important notes.
 *
 * @see https://github.com/conventional-changelog/conventional-changelog/blob/master/packages/conventional-commits-parser/README.md#notekeywords
 */
type NoteKeyword = 'BREAKING CHANGE' | 'BREAKING-CHANGE'

export type { NoteKeyword as default }
