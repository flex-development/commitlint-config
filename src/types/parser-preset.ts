/**
 * @file Type Definitions - ParserPreset
 * @module commitlint-config/types/ParserPreset
 */

import type { ParserOptions } from '#src/interfaces'

/**
 * Object representing a commit parser preset.
 *
 * @see https://commitlint.js.org/#/reference-configuration?id=parser-presets
 */
type ParserPreset = {
  /**
   * Preset name.
   */
  name: 'conventional-changelog-conventionalcommits'

  /**
   * Commit parser options.
   *
   * @see {@linkcode ParserOptions}
   */
  parserOpts: ParserOptions
}

export type { ParserPreset as default }
