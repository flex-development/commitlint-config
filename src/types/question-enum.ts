/**
 * @file Type Definitions - QuestionEnum
 * @module commitlint-config/types/QuestionEnum
 */

/**
 * Object representing prompt question metadata.
 *
 * @see https://commitlint.js.org/#/reference-prompt?id=questions
 */
type QuestionEnum = {
  /**
   * Enum description.
   */
  description?: string

  /**
   * Emoji to display next to {@linkcode title}.
   */
  emoji?: string

  /**
   * Prompt title.
   */
  title?: string
}

export type { QuestionEnum as default }
