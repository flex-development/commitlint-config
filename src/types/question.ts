/**
 * @file Type Definitions - Question
 * @module commitlint-config/types/Question
 */

import type QuestionEnum from './question-enum'

/**
 * Object representing a prompt question.
 *
 * @see https://commitlint.js.org/#/reference-prompt?id=questions
 *
 * @template N - Question enum names
 * @template Q - Question enum type
 */
type Question<
  N extends string = string,
  Q extends QuestionEnum = QuestionEnum
> = {
  /**
   * Question description.
   */
  description: string

  /**
   * Question enums.
   */
  enum?: { [key in N]: Q }

  /**
   * Message hints.
   */
  messages?: { [key in string]: string }
}

export type { Question as default }
