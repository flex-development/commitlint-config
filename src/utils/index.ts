/**
 * @file Entry Point - Utilities
 * @module commitlint-config/utils
 */

export { default as max } from './max'
export { default as min } from './min'
export { default as scopes } from './scopes'
