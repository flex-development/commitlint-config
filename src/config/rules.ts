/**
 * @file Config - rules
 * @module commitlint-config/config/rules
 */

import type { RulesConfig } from '#src/interfaces'
import { max, min, scopes, types } from '#src/utils'
import { RuleConfigSeverity as Severity } from '@commitlint/types'

/**
 * Commit rules.
 *
 * Enforces [conventional commits][1].
 *
 * [1]: https://www.conventionalcommits.org
 *
 * @see {@linkcode RulesConfig}
 * @see https://commitlint.js.org/#/reference-rules
 *
 * @const {RulesConfig} rules
 */
const rules: RulesConfig = {
  'body-case': [Severity.Disabled],
  'body-empty': [Severity.Disabled],
  'body-full-stop': [Severity.Error, 'never', '.'],
  'body-leading-blank': [Severity.Error, 'always'],
  'body-max-length': [Severity.Error, 'always', Number.MAX_SAFE_INTEGER],
  'body-max-line-length': [Severity.Error, 'always', 2050],
  'body-min-length': [Severity.Disabled],
  'footer-empty': [Severity.Disabled],
  'footer-leading-blank': [Severity.Error, 'always'],
  'footer-max-length': [Severity.Error, 'always', Number.MAX_SAFE_INTEGER],
  'footer-max-line-length': [Severity.Error, 'always', 2050],
  'footer-min-length': [Severity.Disabled],
  'header-case': [Severity.Disabled],
  'header-full-stop': [Severity.Error, 'never', '.'],
  'header-max-length': [Severity.Error, 'always', 100],
  'header-min-length': [Severity.Disabled],
  'references-empty': [Severity.Disabled],
  'scope-case': [Severity.Error, 'always', ['kebab-case', 'lower-case']],
  'scope-empty': [Severity.Disabled],
  'scope-enum': [Severity.Error, 'always', scopes()],
  'scope-max-length': [Severity.Error, 'always', max(scopes())],
  'scope-min-length': [Severity.Error, 'always', min(scopes())],
  'signed-off-by': [Severity.Disabled],
  'subject-case': [Severity.Disabled],
  'subject-empty': [Severity.Error, 'never'],
  'subject-exclamation-mark': [Severity.Disabled],
  'subject-full-stop': [Severity.Error, 'never', '.'],
  'subject-max-length': [Severity.Disabled],
  'subject-min-length': [Severity.Error, 'always', 2],
  'trailer-exists': [Severity.Error, 'always', 'Signed-off-by:'],
  'type-case': [Severity.Error, 'always', ['lower-case']],
  'type-empty': [Severity.Error, 'never'],
  'type-enum': [Severity.Error, 'always', types()],
  'type-max-length': [Severity.Error, 'always', max(types())],
  'type-min-length': [Severity.Error, 'always', min(types())]
}

export default rules
