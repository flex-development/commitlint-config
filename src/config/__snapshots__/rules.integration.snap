// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`integration:config/rules > @commitlint/lint > body-full-stop > should error if body ends with "." 1`] = `
[
  {
    "level": 2,
    "message": "body may not end with full stop",
    "name": "body-full-stop",
    "valid": false,
  },
]
`;

exports[`integration:config/rules > @commitlint/lint > body-leading-blank > should error if body comes after 1 new line 1`] = `
[
  {
    "level": 2,
    "message": "body must have leading blank line",
    "name": "body-leading-blank",
    "valid": false,
  },
]
`;

exports[`integration:config/rules > @commitlint/lint > footer-leading-blank > should error if footer comes after 1 new line 1`] = `
[
  {
    "level": 2,
    "message": "footer must have leading blank line",
    "name": "footer-leading-blank",
    "valid": false,
  },
]
`;

exports[`integration:config/rules > @commitlint/lint > header-full-stop > should error if header ends with "." 1`] = `
[
  {
    "level": 2,
    "message": "header may not end with full stop",
    "name": "header-full-stop",
    "valid": false,
  },
  {
    "level": 2,
    "message": "subject may not end with full stop",
    "name": "subject-full-stop",
    "valid": false,
  },
]
`;

exports[`integration:config/rules > @commitlint/lint > header-max-length > should error if header is longer than maximum length 1`] = `
[
  {
    "level": 2,
    "message": "header must not be longer than 100 characters, current length is 133",
    "name": "header-max-length",
    "valid": false,
  },
]
`;

exports[`integration:config/rules > @commitlint/lint > scope-case > should error on invalid scope casing 1`] = `
[
  {
    "level": 2,
    "message": "scope must be kebab-case, lower-case",
    "name": "scope-case",
    "valid": false,
  },
  {
    "level": 2,
    "message": "scope must be one of [build, cjs, cli, components, config, controllers, decorators, deps, deps-bundle, deps-dev, deps-opt, deps-peer, dtos, entities, enums, esm, exceptions, exports, factories, filters, github, guards, hooks, hybrid, install, interceptors, interfaces, internal, lib, loader, middleware, mixins, models, modules, namespaces, node, nvm, patches, pipes, pkg, plugins, ponyfill, props, providers, public, regex, release, rules, scripts, services, spelling, storybook, styles, subdomains, tests, ts, ts-next, types, utils, vercel, vite, vscode, workflows, yarn]",
    "name": "scope-enum",
    "valid": false,
  },
]
`;

exports[`integration:config/rules > @commitlint/lint > scope-enum > should error on invalid scope 1`] = `
[
  {
    "level": 2,
    "message": "scope must be one of [build, cjs, cli, components, config, controllers, decorators, deps, deps-bundle, deps-dev, deps-opt, deps-peer, dtos, entities, enums, esm, exceptions, exports, factories, filters, github, guards, hooks, hybrid, install, interceptors, interfaces, internal, lib, loader, middleware, mixins, models, modules, namespaces, node, nvm, patches, pipes, pkg, plugins, ponyfill, props, providers, public, regex, release, rules, scripts, services, spelling, storybook, styles, subdomains, tests, ts, ts-next, types, utils, vercel, vite, vscode, workflows, yarn]",
    "name": "scope-enum",
    "valid": false,
  },
]
`;

exports[`integration:config/rules > @commitlint/lint > scope-max-length > should error if scope is longer than maximum length 1`] = `
[
  {
    "level": 2,
    "message": "scope must be one of [build, cjs, cli, components, config, controllers, decorators, deps, deps-bundle, deps-dev, deps-opt, deps-peer, dtos, entities, enums, esm, exceptions, exports, factories, filters, github, guards, hooks, hybrid, install, interceptors, interfaces, internal, lib, loader, middleware, mixins, models, modules, namespaces, node, nvm, patches, pipes, pkg, plugins, ponyfill, props, providers, public, regex, release, rules, scripts, services, spelling, storybook, styles, subdomains, tests, ts, ts-next, types, utils, vercel, vite, vscode, workflows, yarn]",
    "name": "scope-enum",
    "valid": false,
  },
  {
    "level": 2,
    "message": "scope must not be longer than 12 characters",
    "name": "scope-max-length",
    "valid": false,
  },
]
`;

exports[`integration:config/rules > @commitlint/lint > scope-min-length > should error if scope is shorter than minimum length 1`] = `
[
  {
    "level": 2,
    "message": "scope must be one of [build, cjs, cli, components, config, controllers, decorators, deps, deps-bundle, deps-dev, deps-opt, deps-peer, dtos, entities, enums, esm, exceptions, exports, factories, filters, github, guards, hooks, hybrid, install, interceptors, interfaces, internal, lib, loader, middleware, mixins, models, modules, namespaces, node, nvm, patches, pipes, pkg, plugins, ponyfill, props, providers, public, regex, release, rules, scripts, services, spelling, storybook, styles, subdomains, tests, ts, ts-next, types, utils, vercel, vite, vscode, workflows, yarn]",
    "name": "scope-enum",
    "valid": false,
  },
  {
    "level": 2,
    "message": "scope must not be shorter than 2 characters",
    "name": "scope-min-length",
    "valid": false,
  },
]
`;

exports[`integration:config/rules > @commitlint/lint > signed-off-by > should error without sign off 1`] = `
[
  {
    "level": 2,
    "message": "message must be signed off",
    "name": "signed-off-by",
    "valid": false,
  },
  {
    "level": 2,
    "message": "message must have \`Signed-off-by:\` trailer",
    "name": "trailer-exists",
    "valid": false,
  },
]
`;

exports[`integration:config/rules > @commitlint/lint > subject-empty > should error on missing subject 1`] = `
[
  {
    "level": 2,
    "message": "subject may not be empty",
    "name": "subject-empty",
    "valid": false,
  },
  {
    "level": 2,
    "message": "type may not be empty",
    "name": "type-empty",
    "valid": false,
  },
]
`;

exports[`integration:config/rules > @commitlint/lint > subject-full-stop > should error if subject ends with "." 1`] = `
[
  {
    "level": 2,
    "message": "header may not end with full stop",
    "name": "header-full-stop",
    "valid": false,
  },
  {
    "level": 2,
    "message": "subject may not end with full stop",
    "name": "subject-full-stop",
    "valid": false,
  },
]
`;

exports[`integration:config/rules > @commitlint/lint > subject-min-length > should error if subject is shorter than minimum length 1`] = `
[
  {
    "level": 2,
    "message": "subject must not be shorter than 2 characters",
    "name": "subject-min-length",
    "valid": false,
  },
]
`;

exports[`integration:config/rules > @commitlint/lint > trailer-exists > should error on missing "Signed-off-by" trailer 1`] = `
[
  {
    "level": 2,
    "message": "message must be signed off",
    "name": "signed-off-by",
    "valid": false,
  },
  {
    "level": 2,
    "message": "message must have \`Signed-off-by:\` trailer",
    "name": "trailer-exists",
    "valid": false,
  },
]
`;

exports[`integration:config/rules > @commitlint/lint > type-case > should error on invalid type casing 1`] = `
[
  {
    "level": 2,
    "message": "type must be lower-case",
    "name": "type-case",
    "valid": false,
  },
  {
    "level": 2,
    "message": "type must be one of [build, chore, ci, docs, feat, fix, perf, refactor, revert, style, test, wip]",
    "name": "type-enum",
    "valid": false,
  },
]
`;

exports[`integration:config/rules > @commitlint/lint > type-empty > should error on missing type 1`] = `
[
  {
    "level": 2,
    "message": "subject may not be empty",
    "name": "subject-empty",
    "valid": false,
  },
  {
    "level": 2,
    "message": "type may not be empty",
    "name": "type-empty",
    "valid": false,
  },
]
`;

exports[`integration:config/rules > @commitlint/lint > type-enum > should error on invalid type 1`] = `
[
  {
    "level": 2,
    "message": "type must be one of [build, chore, ci, docs, feat, fix, perf, refactor, revert, style, test, wip]",
    "name": "type-enum",
    "valid": false,
  },
]
`;

exports[`integration:config/rules > @commitlint/lint > type-max-length > should error if type is longer than maximum length 1`] = `
[
  {
    "level": 2,
    "message": "type must be one of [build, chore, ci, docs, feat, fix, perf, refactor, revert, style, test, wip]",
    "name": "type-enum",
    "valid": false,
  },
  {
    "level": 2,
    "message": "type must not be longer than 8 characters",
    "name": "type-max-length",
    "valid": false,
  },
]
`;

exports[`integration:config/rules > @commitlint/lint > type-min-length > should error if type is shorter than minimum length 1`] = `
[
  {
    "level": 2,
    "message": "type must be one of [build, chore, ci, docs, feat, fix, perf, refactor, revert, style, test, wip]",
    "name": "type-enum",
    "valid": false,
  },
  {
    "level": 2,
    "message": "type must not be shorter than 2 characters",
    "name": "type-min-length",
    "valid": false,
  },
]
`;
